File: sdk/python/captcha-font-sdk/captcha_font_sdk/font_glyph_batch.py
from __future__ import annotations

from pathlib import Path
from typing import Callable, Iterable, List, Optional

from .types import BatchGlyphExtractItem, BatchGlyphExtractResult, FontGlyphFeatureExtractResult


_DEFAULT_IMAGE_EXTS = {".png", ".jpg", ".jpeg", ".webp", ".bmp", ".gif"}


def scan_captcha_files(
    input_dir: str,
    recursive: bool = True,
    exts: Optional[Iterable[str]] = None,
    limit: Optional[int] = None,
) -> List[str]:
    root = Path(input_dir).expanduser().resolve()
    if not root.exists() or not root.is_dir():
        raise FileNotFoundError(f"input_dir not found: {input_dir}")

    valid_exts = _DEFAULT_IMAGE_EXTS if exts is None else {ext.lower() for ext in exts}
    paths = root.rglob("*") if recursive else root.glob("*")
    files = [path for path in paths if path.is_file() and path.suffix.lower() in valid_exts]
    files.sort()
    if limit is not None:
        if limit < 0:
            raise ValueError("limit must be >= 0")
        files = files[:limit]
    return [str(path) for path in files]


def batch_extract_font_glyph_features(
    input_dir: str,
    extractor: Callable[[str], FontGlyphFeatureExtractResult],
    target_width: int,
    target_height: int,
    recursive: bool = True,
    exts: Optional[Iterable[str]] = None,
    limit: Optional[int] = None,
    include_payload: bool = False,
    continue_on_error: bool = True,
) -> BatchGlyphExtractResult:
    files = scan_captcha_files(input_dir, recursive=recursive, exts=exts, limit=limit)

    items: List[BatchGlyphExtractItem] = []
    success_count = 0
    error_count = 0

    for file_path in files:
        try:
            result = extractor(file_path)
            item = BatchGlyphExtractItem(
                captcha_path=file_path,
                status="ok",
                group_id=result.group_id,
                glyph_count=len(result.glyph_features),
                error=None,
                result=result if include_payload else None,
            )
            success_count += 1
            items.append(item)
        except Exception as exc:  # noqa: BLE001
            item = BatchGlyphExtractItem(
                captcha_path=file_path,
                status="error",
                group_id=None,
                glyph_count=0,
                error=str(exc),
                result=None,
            )
            error_count += 1
            items.append(item)
            if not continue_on_error:
                break

    return BatchGlyphExtractResult(
        input_dir=str(Path(input_dir).expanduser().resolve()),
        total_files=len(files),
        processed_files=len(items),
        success_count=success_count,
        error_count=error_count,
        target_size=(target_width, target_height),
        items=items,
    )

File: typescript-sdk/src/font-glyph-batch.ts
import fs from 'fs';
import path from 'path';
import { BatchGlyphExtractItem, BatchGlyphExtractResult, FontGlyphFeatureExtractResult } from './types';

const DEFAULT_IMAGE_EXTS = new Set(['.png', '.jpg', '.jpeg', '.webp', '.bmp', '.gif']);

export async function scanCaptchaFiles(
  inputDir: string,
  recursive: boolean = true,
  exts?: Iterable<string>,
  limit?: number
): Promise<string[]> {
  const root = path.resolve(inputDir);
  if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) {
    throw new Error(`input_dir not found: ${inputDir}`);
  }
  const validExts = exts ? new Set(Array.from(exts).map((e) => e.toLowerCase())) : DEFAULT_IMAGE_EXTS;

  async function scan(dir: string): Promise<string[]> {
    const entries = await fs.promises.readdir(dir, { withFileTypes: true });
    const files: string[] = [];
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (recursive) {
          files.push(...await scan(fullPath));
        }
      } else if (entry.isFile()) {
        if (validExts.has(path.extname(fullPath).toLowerCase())) {
          files.push(fullPath);
        }
      }
    }
    return files;
  }

  let files = await scan(root);
  files.sort();
  if (limit !== undefined) {
    if (limit < 0) {
      throw new Error('limit must be >= 0');
    }
    files = files.slice(0, limit);
  }
  return files;
}

export async function batchExtractFontGlyphFeatures(
  inputDir: string,
  extractor: (captchaPath: string) => Promise<FontGlyphFeatureExtractResult>,
  targetWidth: number,
  targetHeight: number,
  recursive: boolean = true,
  exts?: Iterable<string>,
  limit?: number,
  includePayload: boolean = false,
  continueOnError: boolean = true
): Promise<BatchGlyphExtractResult> {
  const files = await scanCaptchaFiles(inputDir, recursive, exts, limit);

  const items: BatchGlyphExtractItem[] = [];
  let successCount = 0;
  let errorCount = 0;

  for (const filePath of files) {
    try {
      const result = await extractor(filePath);
      items.push({
        captcha_path: filePath,
        status: 'ok',
        group_id: result.group_id,
        glyph_count: result.glyph_features.length,
        error: null,
        result: includePayload ? result : null,
      });
      successCount += 1;
    } catch (error: any) {
      items.push({
        captcha_path: filePath,
        status: 'error',
        group_id: null,
        glyph_count: 0,
        error: error?.message || String(error),
        result: null,
      });
      errorCount += 1;
      if (!continueOnError) {
        break;
      }
    }
  }

  return {
    input_dir: path.resolve(inputDir),
    total_files: files.length,
    processed_files: items.length,
    success_count: successCount,
    error_count: errorCount,
    target_size: [targetWidth, targetHeight],
    items,
  };
}
